### PR TITLE
[core][experimental] Support actor as the InputNode and MultiOutputNode of an ADAG 

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

Currently, ADAG only supports the driver process as the input and output of a DAG. However, if we want to use ADAG as an inference backend and deploy it with Ray Serve, the Ray Serve replica, which is an actor, should be the input and output of the ADAG. This PR supports an actor to be the input and output of an ADAG. That is,

* Driver -> Actor1 -> Driver (supported)
* Actor1 -> Actor2 -> Actor1 (supported by this PR)
* Actor1 -> Actor2 -> Actor3 (not supported)
* Driver -> Actor1 -> Actor2 (not supported)

### Before this PR

The important implementation logic is in `compiled_dag_node.py`.

Without this PR, ADAG always creates a `_driver_actor` on the same node as the driver process. The driver process then reads the output from the shared memory on that node. That is, if we define an ADAG as follows:

```
Driver -> Actor1 -> Driver
```

it actually works as follows under the hood:

```
Driver -> Actor1 -> DAGDriverProxyActor -> Driver
```

### Change 1

This PR checks whether the process compiling the DAG is an actor process or a driver process. If it is a driver process, create a `DAGDriverProxyActor` (`_driver_actor`). If it is an actor process, allocates the actor's actor handle to `self._actor_handle`. When a DAG node is read by a `MultiOutputNode`, we will use `_driver_actor` or `_actor_handle` depending on whether the process is a driver or an actor. 

### Change 2

Without this PR, when we allocate a channel between the an actor and its downstream nodes, we use:

```python
task.output_channel = ray.get(
    fn.remote(
        do_allocate_channel,
        reader_to_node,
        typ=type_hint,
    )
)
```

and then the actor will execute the function `do_allocate_channel`, which will also launch a Ray task in its downstream DAG nodes. The function `_get_reader_node_id` will determine which Ray nodes the downstream DAG nodes are running on.

It will result in a deadlock if the output is an actor. When we try to allocate the channel `Actor 2 -> Actor 1`, we call `ray.get(... do_allocate_channel)` in Actor 1's process. However, Actor 2 also launches a Ray task in Actor 1 to get Actor 1's node ID. Consequently, Actor 1 and Actor 2 end up waiting for each other, causing a deadlock.

```
Actor 1 -> Actor 2 -> Actor 1
```

It doesn't cause a deadlock if the input/output is the driver process. This is because the driver process calls `ray.get(...do_allocate_channel)`, and Actor 1's process launches a Ray task in `DAGDriverProxyActor` to get the node ID.

```
Driver -> Actor 1 -> DAGDriverProxyActor -> Driver
```

* Solution: I get the node ID for each downstream reader handle in the input/output actor and pass the information to the actor.

### Note

I left comments "Change 1" and "Change 2" in this PR. You can focus on reviewing these two parts.

### Possible next steps

* Revisit deadlock detection
* Support Input actor with NCCL channel

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 1]
```diff
@@ -177,7 +153,7 @@ class Channel(ChannelInterface):
     def __init__(
         self,
         writer: Optional[ray.actor.ActorHandle],
-        readers: List[ray.actor.ActorHandle],
+        reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:156]: add comments

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 2]
```diff
@@ -202,19 +178,14 @@ def __init__(
...
-        for reader in readers:
+        assert len(reader_to_node) > 0
+        for reader, _ in reader_to_node:
             assert isinstance(reader, ray.actor.ActorHandle)
 
         if typ is None:
             typ = SharedMemoryType(DEFAULT_MAX_BUFFER_SIZE)
         elif isinstance(typ, int):
             typ = SharedMemoryType(typ)
-        elif not isinstance(typ, SharedMemoryType):
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:213]: This is unreachable

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 3]
```diff
@@ -273,15 +243,9 @@ def __init__(
                 ray.runtime_context.get_runtime_context().get_node_id()
             )
             self._writer_ref = _create_channel_ref(self, typ.buffer_size_bytes)
+            self._reader_node_id = prev_reader_node
 
-            self._reader_node_id = _get_reader_node_id(self, readers[0])
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:277]: This logic is duplicate with the loop above:

```python
            for reader, node in reader_to_node:
                if prev_reader_node is None:
                    prev_reader_node = node
                elif prev_reader_node != node:
                    raise ValueError(
                        f"All reader actors must be on the same node. Actor "
                        f"{prev_reader} is on node {prev_reader_node} while actor "
                        f"{reader} is on node {node}."
                    )
                prev_reader = reader
```

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 4]
```diff
@@ -527,14 +494,14 @@ class CompositeChannel(ChannelInterface):
     def __init__(
         self,
         writer: Optional[ray.actor.ActorHandle],
-        readers: List[ray.actor.ActorHandle],
+        reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:498]: add comments

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 5]
```diff
@@ -35,7 +35,7 @@ class NestedTorchTensorNcclChannel(ChannelInterface):
     def __init__(
         self,
         writer: ray.actor.ActorHandle,
-        readers: List[ray.actor.ActorHandle],
+        reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
Comment (kevin85421) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:38]: add comments

File: python/ray/dag/compiled_dag_node.py [hunk 6]
```diff
@@ -759,6 +787,22 @@ def _preprocess(self) -> None:
             self._input_num_positional_args = max(input_positional_args) + 1
         self._input_kwargs = tuple(input_kwargs)
 
+    def _get_node_id(self, actor_handle: "ray.actor.ActorHandle") -> str:
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:791]: Updated https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/dag/compiled_dag_node.py [hunk 7]
```diff
@@ -53,7 +53,7 @@ class RayDAGArgs(NamedTuple):
 @DeveloperAPI
 def do_allocate_channel(
     self,
-    readers: List[Optional["ray.actor.ActorHandle"]],
+    reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:56]: Updated https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 8]
```diff
@@ -9,10 +9,12 @@
 import sys
 import time
 import numpy as np
+import torch
 
 import pytest
 
 from ray.exceptions import RayChannelError, RayChannelTimeoutError
+import ray.remote_function
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py]: Removed https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/experimental/channel/common.py [hunk 9]
```diff
@@ -76,7 +76,7 @@ def set_contains_type(self, typ: "ChannelOutputType") -> None:
     def create_channel(
         self,
         writer: Optional["ray.actor.ActorHandle"],
-        readers: List[Optional["ray.actor.ActorHandle"]],
+        reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/experimental/channel/common.py:79]: Updated https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 10]
```diff
@@ -123,7 +99,7 @@ def __init__(self, buffer_size_bytes: int):
     def create_channel(
         self,
         writer: Optional["ray.actor.ActorHandle"],
-        readers: List[Optional["ray.actor.ActorHandle"]],
+        reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/experimental/channel/shared_memory_channel.py:102]: Added https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/experimental/channel/shared_memory_channel.py:156]: Added https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 4 above)
Comment (kevin85421) [reply python/ray/experimental/channel/shared_memory_channel.py:498]: Added https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 5 above)
Comment (kevin85421) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:38]: Added https://github.com/ray-project/ray/pull/46700/commits/dfb92861f79f21f51683f324ede9908585c2e3db

File: python/ray/dag/compiled_dag_node.py [hunk 11]
```diff
@@ -759,6 +788,30 @@ def _preprocess(self) -> None:
             self._input_num_positional_args = max(input_positional_args) + 1
         self._input_kwargs = tuple(input_kwargs)
 
+    def _get_node_id(self, actor_handle: "ray.actor.ActorHandle") -> str:
```
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:791]: Change 2

File: python/ray/dag/compiled_dag_node.py [hunk 12]
```diff
@@ -843,32 +896,52 @@ def _get_or_compile(
                     #     print(result)
 
                     #     compiled_dag.teardown()
-                    reader_handles = [self._driver_actor]
+                    assert (self._driver_actor is not None) ^ (
+                        self._actor_handle is not None
+                    )
+                    if self._driver_actor is not None:
+                        reader_to_node.append(
+                            (self._driver_actor, self._get_node_id(self._driver_actor))
```
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:904]: Change 2

File: python/ray/dag/compiled_dag_node.py [hunk 13]
```diff
@@ -843,32 +896,52 @@ def _get_or_compile(
...
             elif isinstance(task.dag_node, InputNode):
+                reader_to_node: List[Tuple["ray.actor.ActorHandle", str]] = []
+                # TODO (kevin85421): We need to remove reader_handles_set because
+                # readers can be on the same actor.
                 reader_handles_set = set()
                 for idx in task.downstream_node_idxs:
                     reader_task = self.idx_to_task[idx]
                     assert isinstance(reader_task.dag_node, ClassMethodNode)
                     reader_handle = reader_task.dag_node._get_actor_handle()
+                    if reader_handle not in reader_to_node:
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:937]: Looks like the type does not match? reader_handle is an actor handle, reader_to_node is a list of tuples

File: python/ray/dag/compiled_dag_node.py [hunk 14]
```diff
@@ -543,11 +545,38 @@ def __init__(
...
+                        "Compiled DAGs currently require the InputNode() to be the "
+                        "driver process or an actor method. Ray task is not supported."
+                    )
+            return None
+
+        self._actor_handle = _get_actor_handle()
+        self._driver_actor = None
+
+        if self._actor_handle is None:
+            self._driver_actor = CompiledDAG.DAGDriverProxyActor.options(
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:575]: Why do we always need this if driver creates the DAG?

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 15]
```diff
@@ -257,8 +229,7 @@ def __init__(
             # different nodes.
             prev_reader_node = None
             prev_reader = None
-            for reader in readers:
-                node = _get_reader_node_id(self, reader)
+            for reader, node in reader_to_node:
```
Comment (ruisearch42) [inline python/ray/experimental/channel/shared_memory_channel.py:232]: Does the logic change anywhere, or passing in `reader_to_node` is just refactoring?

**Review by ruisearch42 — COMMENTED** [review]
First pass review

File: python/ray/dag/compiled_dag_node.py [hunk 16]
```diff
@@ -53,13 +53,14 @@ class RayDAGArgs(NamedTuple):
 @DeveloperAPI
 def do_allocate_channel(
     self,
-    readers: List[Optional["ray.actor.ActorHandle"]],
+    reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:56]: It is not a dict. The list may include the same actor for more than 1 element (multiple readers from the same actor).

File: python/ray/dag/compiled_dag_node.py [hunk 17]
```diff
@@ -521,6 +522,7 @@ def __init__(
         self.actor_to_executable_tasks: Dict[
             "ray.actor.ActorHandle", List["ExecutableTask"]
         ] = {}
+        self.actor_to_node_id: Dict["ray.actor.ActorHandle", List[str]] = {}
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:525]: Good catch. `List[str]` is the wrong type hint. I will change to `str` instead.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 15 above)
Comment (kevin85421) [reply python/ray/experimental/channel/shared_memory_channel.py:232]: You can read "Change 2" in the PR description for more details.

File: python/ray/dag/compiled_dag_node.py (hunk 13 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:937]: Good catch. I initially used a dictionary, but then changed to a list and forgot to update it.

File: python/ray/dag/compiled_dag_node.py (hunk 14 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:575]: This is the existing behavior.

File: python/ray/dag/compiled_dag_node.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Updated https://github.com/ray-project/ray/pull/46700/commits/66e7d5c7436a624690ff8ca3e57e6d4f62333cbd

**Review by stephanie-wang — COMMENTED** [review]
Thanks for this change! It looks good to me overall.

File: python/ray/dag/compiled_dag_node.py [hunk 18]
```diff
@@ -543,11 +546,38 @@ def __init__(
         # create a reader ref for a larger object when the channel backing store is
         # resized, etc.). The `_driver_actor` serves as a way for the output writer to
         # invoke remote functions on the driver node.
-        self._driver_actor = CompiledDAG.DAGDriverProxyActor.options(
-            scheduling_strategy=NodeAffinitySchedulingStrategy(
-                ray.get_runtime_context().get_node_id(), soft=False
-            )
-        ).remote()
+
+        def _get_actor_handle():
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:550]: Good point!

> Suggest rolling the logic to create the proxy actor into this helper function and moving it to util.py.

Do you refer to this file https://github.com/ray-project/ray/blob/master/python/ray/dag/utils.py?

>  I think the main thing is that we want to make sure there is an actor process local to the DAG's creator, right?

Yes

File: python/ray/dag/compiled_dag_node.py [hunk 19]
```diff
@@ -808,22 +859,21 @@ def _get_or_compile(
...
                             "DAG outputs currently can only be read by the driver--not "
                             "the driver and actors."
                         )
-                    # This node is a multi-output node, which means that it will only be
-                    # read by the driver, not an actor. Thus, we handle this case by
-                    # setting `reader_handles` to `[self._driver_actor]`.
+                    # This node is a multi-output node, which means it will only be
+                    # read by the driver or the actor that is also the InputNode.
 
                     # TODO(jhumphri): Handle case where there is an actor, other than
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:878]: is this comment still relevant?

File: python/ray/dag/compiled_dag_node.py [hunk 20]
```diff
@@ -843,32 +893,45 @@ def _get_or_compile(
...
-                        reader_handles,
+                        reader_to_node,
                         typ=type_hint,
                     )
                 )
                 actor_handle = task.dag_node._get_actor_handle()
                 self.actor_refs.add(actor_handle)
                 self.actor_to_tasks[actor_handle].append(task)
             elif isinstance(task.dag_node, InputNode):
+                reader_to_node: List[Tuple["ray.actor.ActorHandle", str]] = []
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:919]:
```suggestion
                reader_to_node_id: List[Tuple["ray.actor.ActorHandle", str]] = []
```

File: python/ray/experimental/channel/common.py [hunk 21]
```diff
@@ -76,16 +76,16 @@ def set_contains_type(self, typ: "ChannelOutputType") -> None:
...
```
Comment (rkooo567) [inline python/ray/experimental/channel/common.py:79]:
```suggestion
        readers_to_nodes: List[Tuple["ray.actor.ActorHandle", str]],
```

**Review by rkooo567 — APPROVED** [review]
Looks good to me

File: python/ray/dag/compiled_dag_node.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: > Suggest rolling the logic to create the proxy actor into this helper function and moving it to util.py.

The function seems only to be used in `CompiledDAG`'s constructor. Do we still need to move to `util.py`?

File: python/ray/dag/compiled_dag_node.py (hunk 19 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:878]: Yes, it is still relevant. The comment applies to the case where a DAG node is read by both the driver process and another actor. For example, `a.inc` is read by actor b and the driver process at the same time. This example doesn't work currently.

```
driver -> a.inc -> b.inc -> driver 
                |        |
                ----------
```

File: python/ray/experimental/channel/common.py (hunk 21 above)
Comment (kevin85421) [reply python/ray/experimental/channel/common.py:79]: How about changing all `reader_to_node` to `reader_to_node_id` instead?

File: python/ray/dag/compiled_dag_node.py [hunk 22]
```diff
@@ -843,32 +893,45 @@ def _get_or_compile(
...
+                # TODO (kevin85421): We need to remove reader_handles_set because
+                # readers can be on the same actor.
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:921]: Can you make the TODO clearer? e.g., adding information why this is no longer a constraint

**Review by ruisearch42 — COMMENTED** [review]
The approach looks good. Some suggestions for improving readability.

File: python/ray/dag/compiled_dag_node.py [hunk 23]
```diff
@@ -53,13 +53,14 @@ class RayDAGArgs(NamedTuple):
...
-    readers: List[Optional["ray.actor.ActorHandle"]],
+    reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
     typ: ChannelOutputType,
 ) -> ChannelInterface:
     """Generic actor method to allocate an output channel.
 
     Args:
-        readers: The actor handles of the readers.
+        reader_to_node: A list of tuples, where each tuple contains a reader
+            actor handle and the node ID where the handle is located.
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Update https://github.com/ray-project/ray/pull/46700/commits/3a9b6b8a9833a4d46c5dbba8b28567d6197b0955

File: python/ray/dag/compiled_dag_node.py [hunk 24]
```diff
@@ -53,13 +53,14 @@ class RayDAGArgs(NamedTuple):
...
     self,
-    readers: List[Optional["ray.actor.ActorHandle"]],
+    reader_to_node: List[Tuple["ray.actor.ActorHandle", str]],
     typ: ChannelOutputType,
 ) -> ChannelInterface:
     """Generic actor method to allocate an output channel.
 
     Args:
-        readers: The actor handles of the readers.
+        reader_to_node: A list of tuples, where each tuple contains a reader
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Updated https://github.com/ray-project/ray/pull/46700/commits/3a9b6b8a9833a4d46c5dbba8b28567d6197b0955

File: python/ray/dag/compiled_dag_node.py [hunk 25]
```diff
@@ -808,22 +859,21 @@ def _get_or_compile(
                 # `readers` is the nodes that are ordered after the current one (`task`)
                 # in the DAG.
                 readers = [self.idx_to_task[idx] for idx in task.downstream_node_idxs]
-
+                reader_to_node: List[Tuple["ray.actor.ActorHandle", str]] = []
                 dag_nodes = [reader.dag_node for reader in readers]
-                read_by_driver = False
+                read_by_output_node = False
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: I think the name is fine, but I still update the name to `read_by_multi_output_node`.

File: python/ray/dag/compiled_dag_node.py [hunk 26]
```diff
@@ -808,22 +859,21 @@ def _get_or_compile(
...
                     if isinstance(dag_node, MultiOutputNode):
-                        read_by_driver = True
+                        read_by_output_node = True
                         break
-                if read_by_driver:
+                if read_by_output_node:
                     if len(readers) != 1:
                         raise ValueError(
                             "DAG outputs currently can only be read by the driver--not "
                             "the driver and actors."
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Updated https://github.com/ray-project/ray/pull/46700/commits/3a9b6b8a9833a4d46c5dbba8b28567d6197b0955

File: python/ray/dag/compiled_dag_node.py [hunk 27]
```diff
@@ -536,18 +539,42 @@ def __init__(
...
+            # create a reader ref for a larger object when the channel backing store is
+            # resized, etc.). The driver actor serves as a way for the output writer
+            # to invoke remote functions on the driver node.
+            return CompiledDAG.DAGDriverProxyActor.options(
+                scheduling_strategy=NodeAffinitySchedulingStrategy(
+                    ray.get_runtime_context().get_node_id(), soft=False
+                )
+            ).remote()
+
+        self._actor_handle = _get_or_create_local_actor_handle()
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Updated https://github.com/ray-project/ray/pull/46700/commits/3a9b6b8a9833a4d46c5dbba8b28567d6197b0955

File: python/ray/dag/compiled_dag_node.py (hunk 22 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: Updated https://github.com/ray-project/ray/pull/46700/commits/3a9b6b8a9833a4d46c5dbba8b28567d6197b0955

**Pull request merged**